src/sentry/taskworker/retry.py

```diff
@@ -0,0 +1,64 @@
+from __future__ import annotations
+
+from collections.abc import Sequence
+from multiprocessing.context import TimeoutError
+
+from sentry_protos.sentry.v1.taskworker_pb2 import RetryState
+
+
+class Retry:
+    __times: int
+    __allowed_exception_types: Sequence[type] | None
+    __denied_exception_types: Sequence[type] | None
+    __deadletter: bool | None
+    __discard: bool | None
+
+    """Used with tasks to define the retry policy for a task"""
+
+    def __init__(
+        self,
+        times: int = 1,
+        on: Sequence[type] | None = None,
+        ignore: Sequence[type] | None = None,
+        deadletter: bool | None = None,
+        discard: bool | None = None,
+    ):
+        self.__times = times
+        self.__allowed_exception_types = on
+        self.__denied_exception_types = ignore
+        self.__deadletter = deadletter
+        self.__discard = discard
+
+    def should_retry(self, state: RetryState, exc: Exception) -> bool:
+        # No more attempts left
+        if state.attempts >= self.__times:
+            return False
+        # No retries for types on the ignore list
+        if any(
+            isinstance(exc, ignored_exception_type)
+            for ignored_exception_type in self.__denied_exception_types or []
+        ):
+            return False
+        # In the retry allow list or processing deadline is exceeded
+        # When processing deadline is exceeded, the subprocess raises a TimeoutError
+        if isinstance(exc, TimeoutError) or (
+            any(
+                isinstance(exc, allowed_exception_type)
+                for allowed_exception_type in self.__allowed_exception_types or []
+            )
+        ):
+            return True
```

Comment:
There were some typing problems here and I think we were not checking the list of types properly. Renamed to make it more clear and check types using `any`